#8: Implemented a release workflow

tomuben opened the pull request:
fixes #8

tomuben committed:
moved ci_codebuild to setup_ci_codebuild

tomuben committed:
updated changelog

tomuben committed:
added start_codebuild to AwsAccess

tomuben committed:
1. added the start_release_build and start_test_release_build
2. Added setup_release_codebuild
3. added update_release

tomuben committed:
updated poetry version in gh actions

tomuben committed:
updated poetry version in aws codebuild buildspecs

tomuben committed:
Update setup

tomuben committed:
updated docstrings

tomuben committed:
updated test_deploy_codebuild.py such that it tests also release codebuild cloudformation template

tomuben committed:
updated curl version in ansible scripts

tomuben committed:
added newline in __init__.py

tomuben committed:
added test_release_build.py

tomuben committed:
updated gh workflows

tkilias commented on exasol_script_languages_developer_sandbox/lib/github_release_access.py:
```diff
@@ -0,0 +1,60 @@
+        """
+        release = self._get_repo.get_release(release_id)
+        body = release.body
+        body += message_to_append
```
Maybe to be safe, add a line break

tomuben replied:
Done

tkilias commented on exasol_script_languages_developer_sandbox/lib/github_release_access.py:
```diff
@@ -0,0 +1,60 @@
+        body += message_to_append
+        release.update_release(release.title, body)
+
+    def upload(self, archive_path: str, label: str, release_id: int, content_type: str):
```
Do we have assets to upload? Maybe the checksums of the inages

tomuben replied:
I upload currently the links of the assets as markdown file

tkilias replied:
Ok

tkilias commented on exasol_script_languages_developer_sandbox/lib/release_build/run_release_build.py:
```diff
@@ -0,0 +1,79 @@
+                     ("ASSET_ID", f"{asset_id}"),
+                     ("GITHUB_TOKEN", gh_token)]
+    environment_variables_overrides = list(map(get_environment_variable_override, env_variables))
+    aws_access.start_codebuild(matching_project.physical_id,
```
Why not put a dict into start_codebuold for env_variables and let start_codebuold do the transformation

tomuben replied:
I want to keep the implementation in AwsAccess as thin as possible, just a wrapper for boto3, so that I can test the rest.
(Although I have to admit that start_codebuild() did not follow really this guideline).

tkilias commented on test/test_release_build.py:
```diff
@@ -0,0 +1,79 @@
+    """
+    Test that serialization and deserialization of KeyFileManager work!
+    """
+    aws_access_mock = create_autospec(AwsAccess, spec_set=True)
```
Maybe use union mock type hints and mock cast

tomuben replied:
Done

tomuben committed:
Fixed buildspec_release.yaml

tomuben committed:
fixed AwsAccess.start_codebuild()

tomuben committed:
fixed poetry path in AWS Codebuild buildspecs

tomuben committed:
changed retrieval of user name (which is used for AWS resource tagging only!):
1. IAM roles cannot access the profile user name
2. For that, internally we pass an optional user_name parameter
3. This parameter can be set via envrionment variable only, not command parameter.

tomuben committed:
Made release notes more beautiful

tomuben committed:
Fixed permission for IAM role

tomuben committed:
replace also whitespace for stack-prefix

tomuben committed:
changed permission for IAM role in release_code_build.jinja.yaml

tomuben committed:
extended documentation for AwsAccess.start_codebuild()

tomuben committed:
use mock_cast in test_release_build.py

tomuben committed:
updated permissions for IAM role in release_code_build.jinja.yaml

tomuben committed:
Let run_export_vm fail if any of the exports fails.
(Otherwise AWS build will be marked as "succeeded")

tomuben committed:
Updated permission in release_code_build.jinja.yaml

tomuben committed:
fixed tests

tomuben committed:
fixed test_ci.py

tomuben committed:
1. encode http url in print_assets.py correctly.
2. tiny change in mark_down_printer.py

tomuben committed:
reverted usage of master branch for ci-builds

tomuben committed:
[CodeBuild]

tomuben committed:
fixed documentation in aws_access.py

tomuben committed:
[CodeBuild]

tkilias commented on exasol_script_languages_developer_sandbox/lib/setup_ec2/run_setup_ec2.py:
```diff
@@ -21,9 +21,19 @@
 EC2LifecycleDataIterator = Iterator[EC2LifecycleData]
 
 
+def retrieve_user_name(user_name: Optional[str], aws_access: AwsAccess) -> str:
```
Add reason why we need this

tomuben replied:
Done

tkilias commented on exasol_script_languages_developer_sandbox/lib/github_release_access.py:
```diff
@@ -31,8 +31,8 @@ def update_release_message(self, release_id: int, message_to_append: str) -> Non
-        body += message_to_append
-        release.update_release(release.title, body)
+        body += "\n\n" + message_to_append
+        release.update_release(name=release.title, message=body, draft=True, prerelease=True)
```
Do we need prerelease if we do draft

tomuben replied:
Hm, the docu says: "To notify users that the release is not ready for production and may be unstable, select This is a pre-release." So, I think you are right, that's not our case.

tomuben replied:
Done

tkilias commented on exasol_script_languages_developer_sandbox/templates/release_code_build.jinja.yaml:
```diff
@@ -32,7 +32,7 @@ Resources:
                 - cloudformation:ExecuteChangeSet
                 - cloudformation:DeleteStack
-                Resource: !Sub  "arn:aws:cloudformation:${AWS::Region}:${AWS::AccountId}:stack/stack-ci-test-*"
+                Resource: !Sub  "arn:aws:cloudformation:${AWS::Region}:${AWS::AccountId}:stack/*"
```
I guess this is because of the release code build stack. Can we make them share a common prefix, because this way , it gets permissions over all stacks and we might have also other stacks in the account

tomuben replied:
Was already fixed in a later commit. Please check again

tkilias replied:
ah, ok I thought so, but wasn't sure, because I went commit by commi

tkilias commented on exasol_script_languages_developer_sandbox/lib/aws_access/aws_access.py:
```diff
@@ -337,10 +337,10 @@ def get_user(self) -> str:
-        After the build has triggered it waits until the batch build finished
+        It starts the codebuild for the given branch.
+        After the build has triggered it waits until the build finished.
         :param project: Codebuild project name to start
```
After  the build was triggered, it waits

tomuben replied:
done

tomuben committed:
fixes from reviews

tomuben committed:
[CodeBuild]

tomuben commented on exasol_script_languages_developer_sandbox/templates/release_code_build.jinja.yaml:
```diff
@@ -32,7 +32,7 @@ Resources:
                 - cloudformation:ExecuteChangeSet
                 - cloudformation:DeleteStack
-                Resource: !Sub  "arn:aws:cloudformation:${AWS::Region}:${AWS::AccountId}:stack/*"
+                Resource: !Sub  "arn:aws:cloudformation:${AWS::Region}:${AWS::AccountId}:stack/EC2-SLC-DEV-SANDBOX-*"
```
here it's updated with the prefix

tomuben committed:
updated docstring in aws_access.py

tomuben committed:
[CodeBuild]

tomuben committed:
[CodeBuild]
